ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/SignatureVerificationService.java

```diff
@@ -0,0 +1,162 @@
+import tech.pegasys.teku.spec.logic.common.util.AsyncBLSSignatureVerifier;
+
+public class SignatureVerificationService extends Service implements AsyncBLSSignatureVerifier {
+  private static final int QUEUE_CAPACITY = 1000;
+  private static final int MAX_BATCH_SIZE = 50;
+  private static final Duration INACTIVITY_PAUSE = Duration.ofMillis(200);
+
+  private final ExecutorFactory executorFactory;
+  private final int numThreads;
+  private final int maxBatchSize;
+
+  private final BlockingQueue<SignatureTask> batchSignatureTasks;
+  private ScheduledExecutorService executor;
+
+  @VisibleForTesting
+  SignatureVerificationService(
+      final ExecutorFactory executorFactory,
+      final int numThreads,
+      final int queueCapacity,
+      final int maxBatchSize) {
+    this.executorFactory = executorFactory;
+    this.numThreads = Math.min(numThreads, Runtime.getRuntime().availableProcessors());
+    this.maxBatchSize = maxBatchSize;
+
+    this.batchSignatureTasks = new ArrayBlockingQueue<SignatureTask>(queueCapacity);
+  }
+
+  public static SignatureVerificationService create() {
+    return new SignatureVerificationService(
+        SignatureVerificationService::defaultExecutorFactory, 2, QUEUE_CAPACITY, MAX_BATCH_SIZE);
+  }
+
+  private static ScheduledExecutorService defaultExecutorFactory(final int numThreads) {
+    return new ScheduledThreadPoolExecutor(
+        numThreads,
+        new ThreadFactoryBuilder()
+            .setDaemon(false)
+            .setNameFormat(SignatureVerificationService.class.getSimpleName() + "-%d")
+            .build());
+  }
+
+  @Override
+  protected synchronized SafeFuture<?> doStart() {
+    executor = executorFactory.createExecutor(numThreads);
+    for (int i = 0; i < numThreads; i++) {
+      executor.submit(this::run);
+    }
+
+    return SafeFuture.COMPLETE;
+  }
+
+  @Override
+  protected synchronized SafeFuture<?> doStop() {
+    executor.shutdownNow();
+    return SafeFuture.COMPLETE;
+  }
+
+  @Override
+  public SafeFuture<Boolean> verify(
+      final List<BLSPublicKey> publicKeys, final Bytes message, final BLSSignature signature) {
+    assertIsRunning("verify");
+    final SignatureTask task = new SignatureTask(publicKeys, message, signature);
+    if (!batchSignatureTasks.offer(task)) {
+      // Queue is full
+      final Throwable error =
+          new ServiceCapacityExceededException("Failed to process signature, queue is full.");
+      task.result.completeExceptionally(error);
+    }
+    return task.result;
+  }
+
+  private void run() {
+    while (isRunning()) {
+      final List<SignatureTask> tasks = new ArrayList<>();
+      batchSignatureTasks.drainTo(tasks, maxBatchSize);
+      if (tasks.isEmpty()) {
+        // Exit loop and resume after a short pause
+        executor.schedule(this::run, INACTIVITY_PAUSE.toMillis(), TimeUnit.MILLISECONDS);
+        return;
+      } else {
+        batchVerifySignatures(tasks);
+      }
```

Comment:
I'd be inclined to block on the queue rather than having a fixed delay on inactivity.  So something like:

```
  private void run() {
    while (isRunning()) {
      final List<SignatureTask> tasks = waitForBatch();
      if (!tasks.isEmpty()) {
        batchVerifySignatures(tasks);
      }
    }
  }

  private List<SignatureTask> waitForBatch() {
    final List<SignatureTask> tasks = new ArrayList<>();
    try {
      final SignatureTask firstTask = batchSignatureTasks.poll(30, TimeUnit.SECONDS);
      tasks.add(firstTask);
      batchSignatureTasks.drainTo(tasks, maxBatchSize - 1);
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
    }
    return tasks;
  }
```

The advantage here is that if we're keeping up, we'll process signatures as fast as possible, but if we fall behind or get a sudden spike in the number of signatures we'll start processing in batches. That gives us no maximum throughput with no additional latency.  The timeout on the poll is just to avoid getting stuck permanently - just got to make sure we do interrupt the thread on shutdown so it doesn't have to wait for it to exit.